### ISSUE TITLE
Can not use CombinedReducer in the same way as in ReduxJS

### ISSUE THREAD

**Issue by tuyencoder** [ReSwift/ReSwift#80]
My app has an AppState with 4 substates: SearchScopeState, CloudState, AccountState, FileState. I also created one AppReducer with 4 other reducers SearchScopeReducer, CloudReducer, AccountReducer, FileReducer.

Here is my AppReducer

```swift
struct AppReducer: Reducer {
    func handleAction(action: Action, state: AppState?) -> AppState {
        var state = state ?? AppState()
        state.searchScopeState = SearchScopeReducer().handleAction(action, state: state.searchScopeState)
        state.cloudState = CloudReducer().handleAction(action, state: state.cloudState)
        state.accountState = AccountReducer().handleAction(action, state: state.accountState)
        state.fileState = FileReducer().handleAction(action, state: state.fileState)
        return state
    }
}
```

Here is my store

```swift
let mainStore = Store(reducer: AppReducer(), state: AppState(), middleware: [loggingMiddleware])
```

Now I want to use a CombinedReducer of 4 child reducers. This way I can remove AppReducer, let CombinedReducer do its job. Here is new setup

```swift
let combinedReducer = CombinedReducer([SearchScopeReducer(), CloudReducer(), AccountReducer(), FileReducer()])
let mainStore = Store(reducer: combinedReducer, state: AppState(), middleware: [loggingMiddleware])
```

But when I run the app, the handleAction method of each reducer in the new setup would not have been called. While from the Redux documentation, I can understand that CombinedReducer can help to consolidate substates into one state.

P/S: I have just stated getting familiar with Redux and ReSwift

**Comment by Ben-G** [comment]
Hey @tuyencoder. Sorry about the confusion. `CombinedReducer` is currently relying on some older ideas of the framework. And I'm not entirely sure if we should keep them around.

When you use `CombinedReducer` all of your reducers will receive the top-level app state. That's the case because we didn't have a way to subselect a state they way Redux can do it (due to the static nature of Swift).

Also, your reducers will currently only be called if their expected types, match the types that are actually being passed into them (I'm planning to fix this in #4).

**The Solution:**

1. You can either change the `handleAction` method of your reducer to take `AppState?` as argument and return `AppState`
2. Or, you can define a protocol that allows to access a specific substate then your `handleAction` method can take that substate (e.g. `FileState`). You can read more about that in the "Viewing the State Through a Protocol" section: http://reswift.github.io/ReSwift/master/getting-started-guide.html

**For now I would recommend not using `CombinedReducer` and working with your own solution as shown above.** We should know pretty soon wether we can offer an improved version of `CombinedReducer` or wether we should remove it entirely.

Sorry for the confusion!

**Comment by Ben-G** [comment]
Seems resolved! If not we can reopen.
